File: leetcode/141.Linked-List-Cycle.py
"""
https://leetcode.com/problems/linked-list-cycle/description/

Time complexity: O(n), n = len(head)
Space complexity: O(1)
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head == None: return False
        fast, slow = head.next, head
        while fast and fast.next:
            if fast == slow: return True
            fast = fast.next.next
            slow = slow.next
        return False

File: engineering/LeetCode/141.linked-list-cycle.ts
/*
 * @lc app=leetcode id=141 lang=typescript
 *
 * [141] Linked List Cycle
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function hasCycle(head: ListNode | null): boolean {
    if (head === null || head.next === null) { return false; }

    let fast = head, slow = head.next;
    while (fast !== null && slow !== null && slow.next !== null) {
        fast = fast.next;
        slow = slow.next.next;

        if (fast === slow) { return true; }
    }

    return false;
};
// @lc code=end

